### ISSUE TITLE
I don't like how I wrote this library

### ISSUE THREAD

**Issue by mileslucas** [JuliaAstro/PSFModels.jl#5]
I'm working on rewriting this library a bit, because I don't really like it. I think I've overdone the complexity with the `PSFModel` type and sub-typing `AbstractArray`. I've found the array approach works well when you just want to plot a single PSF, but it's not really useful when *fitting* PSF models, which is how I actually use this library. 

I'm going to try and create a simpler interface, that uses functions more alike
```
gaussian([T], px, py; x, y, fwhm, amp=1, theta=0)
```
instead of structs masquerading as functions. This means that a single evaluation will always be allocation-free, and iterating over many points is effortless

```julia
img = randn(1:10, 1:10)
model(idx) = gaussian(idx.I...; x=5.5, y=5.5, fwhm=(2, 3))
loss = sum(CartesianIndices(img)) do idx
	val = model(idx)
    return (val - img[idx])^2
end
loss /= length(img)
```

with the appropriate methods, we can dispatch on `CartesianIndex` or whatever, too. We can write out curried versions, too, which essentially fully recreates the existing API, just way simpler.

I'm unsure of supporting `(r, theta)` syntax anymore for the PSFs, since I've never really found it useful, and I don't use it in the packages which depend on this.

This may also simplify the implementation of ChainRules, since I've had nothing but trouble trying to get those working on structs that are essentially just curried functions (for example, I have to manually define the gradient w.r.t. the indices of the pseudo-array in the current library).

I also want to implement some fitting methods. I'm unsure whether they should be in this repository or another one. For now I think I'll just implement it here (along with the dependency for `LossFunctions.jl`/`Optim.jl` or whichever fitting package I use). In the future perhaps there's a split into `PSFFunctions.jl` and `PSFModels.jl` or perhaps `PSFModels.jl` and `PSFFitters.jl`. That will be the case for packages which want PSFs without the "bloat" of the fitting packages.
